static/js/filter.js

```diff
@@ -433,6 +434,147 @@ Filter.prototype = {
+    // common narrows show a narrow description and allow the user to
+    // close search bar UI and show the narrow description UI.
+    is_common_narrow: function () {
+        // can_mark_messages_read tests the following filters:
+        // stream, stream + topic,
+        // is: private, pm-with:,
+        // is: mentioned
+        if (this.can_mark_messages_read()) {
+            return true;
+        }
+        // that leaves us with checking:
+        // is: starred
+        // (which can_mark_messages_read_does not check as starred messages are always read)
+        const term_types = this.sorted_term_types();
+
+        if (_.isEqual(term_types, ['is-starred'])) {
+            return true;
+        }
+    },
+
+    // This is used to control the behaviour for "exiting search",
+    // for narrows where there is custom logic as per
+    // https://paper.dropbox.com/doc/Navbar-behavior-table--AvnMKN4ogj3k2YF5jTbOiVv_AQ-cNOGtu7kSdtnKBizKXJge
+    // all others redirect to "All"
+    generate_redirect_url: function () {
+        const term_types = this.sorted_term_types();
+
+        // this comes first because it has 3 term_types but is not a "complex filter"
+        if (_.isEqual(term_types, ['stream', 'topic', 'search'])) {
+            return  '/#narrow/stream/' + stream_data.name_to_slug(this.operands('stream')[0]) + '/topic/' + this.operands('topic')[0];
+        }
+
+        // eliminate "complex filters"
+        if (term_types.length >= 3) {
+            return "#"; // redirect to All
+        }
+
+        if (term_types[1] === 'search') {
+            switch (term_types[0]) {
+            case 'stream':
+                return  '/#narrow/stream/' + stream_data.name_to_slug(this.operands('stream')[0]);
+            case 'is-private':
+                return  '/#narrow/is/private';
+            case 'is-starred':
+                return  '/#narrow/is/starred';
+            case 'is-mentioned':
+                return  '/#narrow/is/mentioned';
+            case 'pm-with':
+                // join is used to transform the array to a comma separated string
+                return  '/#narrow/pm-with/' + people.emails_to_slug(this.operands('pm-with').join());
+                // TODO: It is ambiguous how we want to handle the 'sender' case,
+                // we may remove it in the future based on design decisions
+            case 'sender':
+                return  '/#narrow/sender/' + people.emails_to_slug(this.operands('sender')[0]);
+            }
+        }
+
+        return "#"; // redirect to All
+    },
+
+    get_icon: function () {
+        const term_types = this.sorted_term_types();
+        if (term_types.length === 1 && ['in-home', 'in-all'].includes(term_types[0])) {return 'home';}
+        switch (term_types[0]) {
+        case 'stream':
+            if (stream_data.get_sub_by_name(this.operands('stream')[0]).invite_only) {
```

Comment:
You end up doing this `stream_data` lookup a million times.  Would it make sense to do this lookup just at initialization time and store the result on the `Filter` object?  Seems a lot better than writing this `get_sub_by_name` thing 15 times.